update snowflake private-key handling to work with uploaded private keys

escherize opened the pull request:
Resolves #26078

The snowflake pk can come in a number of ways: 

- an uploaded string
- an uploaded byte array
- a local path to a file with the key

We should be able to connect using a private key from any of those methods.

escherize committed:
treat private-key conn props as base64 encoded

- fix typo

escherize committed:
handle connecting with local keypair

escherize committed:
adds test for snowflake keypair auth

escherize committed:
remove unused arg

escherize commented on modules/drivers/snowflake/src/metabase/driver/snowflake.clj:
```diff
@@ -326,13 +326,45 @@
+       (some? account)
+       (some? user)))
+
+(let [ByteArray (type (byte-array []))]
```
I did it this way because

```clojure
(type (byte-array []))
;;=> [B
```

and `[B` is not readable. Is there a better way to do this?

escherize committed:
remove id, and only call byte-array->string once

escherize committed:
dont reflect for conversion from byte-array to str

escherize committed:
remove inline def

escherize committed:
more cleanup

- mess with tests to find the fabled metabase ci snowflake user

escherize committed:
Merge branch 'master' into fix-snowflake-private-key-issue

escherize committed:
remove un-needed test

escherize committed:
handle conn uri in connection-details->spec

escherize committed:
Merge branch 'master' into fix-snowflake-private-key-issue

escherize committed:
Merge branch 'fix-snowflake-private-key-issue' of github.com:metabase/metabase into fix-snowflake-private-key-issue

escherize committed:
make sure tests are actually running

escherize committed:
encode properties in the connection uri

escherize committed:
typehint private keyfile

tsmacdonald commented on modules/drivers/snowflake/test/metabase/driver/snowflake_test.clj:
```diff
@@ -123,27 +124,34 @@
+               net.snowflake.client.jdbc.SnowflakeSQLException
+               (can-connect? (assoc (:details (mt/db)) :db (mt/random-name))))
+              "can-connect? should throw for Snowflake databases that don't exist (#9511)")
+          (when (and pk-key pk-user)
```
When would the key and user env vars not be set?

escherize replied:
Those will be nil whenever the env variables `MB_SNOWFLAKE_TEST_PK_PRIVATE_KEY` and `MB_SNOWFLAKE_TEST_PK_USER` are not set

escherize replied:
So, if you were to run snowflake tests without wanting to test the key pair auth (which requires a lot of setup), then this test will still pass

metamben commented on modules/drivers/snowflake/test/metabase/driver/snowflake_test.clj:
```diff
@@ -123,27 +124,34 @@
+               (can-connect? (assoc (:details (mt/db)) :db (mt/random-name))))
+              "can-connect? should throw for Snowflake databases that don't exist (#9511)")
+          (when (and pk-key pk-user)
+            (log/fatal (str/join (repeat 1000 "[TEST DEBUG MESSAGE] - Running private key authentication Test\n")))
```
This should be removed, I suppose.

escherize commented on modules/drivers/snowflake/src/metabase/driver/snowflake.clj:
```diff
@@ -54,19 +56,37 @@
+         (format "jdbc:snowflake://%s.snowflakecomputing.com?user=%s&private_key_file=%s"
+                 account
+                 (codec/url-encode user)
+                 (codec/url-encode (.getCanonicalPath ^File private-key-file)))))
```
Like what exactly? The users don't get to set their connection-uri, and we don't have anything else that adds query params to it for snowflake, right?

metamben replied:
That might be right at the moment. It still disturbs me somewhat that we are setting a global thing that might conflict with other parameters we might want to set via properties or with another place where we might want to set the URI.

escherize replied:
So, if someone can set it via a property or something, then we should keep their settings, and merge in the necessary fields. That makes sense. I will add that piece

escherize commented on modules/drivers/snowflake/src/metabase/driver/snowflake.clj:
```diff
@@ -330,9 +350,8 @@
-           (jdbc/query spec sql)
-           true))))
+         (jdbc/query spec (format "SHOW OBJECTS IN DATABASE \"%s\";" db)))
+       true))
```
I think so. I guess before it was just seeing if the query would throw. It's probably safer to adhere to the original assumption, so I'll explicitly wrap it. That way we can let users connect to an empty database.

escherize replied:
I added a test for making sure can-connect? will return true, even when `jdbc/query` returns as falsey

escherize committed:
respond to some review comments

- add a test for can-connect? => true, even when jdbc/query returns falsey

escherize committed:
remove unused logging require

escherize committed:
Merge remote-tracking branch 'origin/master' into fix-snowflake-private-key-issue

escherize committed:
adds assoc-qp

- which will be used to non-destructively "assoc" onto the snowflake
connection query params

escherize committed:
hook in u.qp/assoc-qp to better update conn uri

escherize committed:
kill a newline

escherize committed:
Merge branch 'master' of github.com:metabase/metabase into fix-snowflake-private-key-issue

metamben commented on src/metabase/util/query_params.clj:
```diff
@@ -0,0 +1,38 @@
+                (str url-key "=" (codec/url-encode qp-v))))))
+     new-url
+     ;; sort qps to keep equality stable:
+     (sort qps))))
```
There are a bunch of functions in `metabase.driver.sql-jdbc.common` related to handling additional options which are options encoded in the connection string. You might be able to reuse those or otherwise consolidate that code with yours.

escherize replied:
Fair enough, [here's](https://github.com/metabase/metabase/pull/26304/commits/4c4ea30462e95998c7c6586e1b7e24025c01fd3f#diff-5472d9d00fa628881a6887a87e2be860f2d679f85aeada5d9f2c86173646800aR60-R65) the updated version that resuses `conn-str-with-additional-opts`.

escherize committed:
reuse `conn-str-with-additional-opts` on snowflake conn uri

escherize committed:
Merge branch 'master' into fix-snowflake-private-key-issue

metamben commented on modules/drivers/snowflake/src/metabase/driver/snowflake.clj:
```diff
@@ -54,19 +56,38 @@
+        private-key-file (handle-conn-uri user account private-key-file)))
+
+    private-key-value
+    (let [private-key-str  (if (instance? (Class/forName "[B") private-key-value)
```
```suggestion
    (let [private-key-str  (if (bytes? private-key-value)
```

metamben commented on modules/drivers/snowflake/src/metabase/driver/snowflake.clj:
```diff
@@ -54,19 +56,38 @@
+
+    private-key-value
+    (let [private-key-str  (if (instance? (Class/forName "[B") private-key-value)
+                             (String. ^"[B" private-key-value)
```
Is using the platform's default charset the best option?

escherize commented on modules/drivers/snowflake/src/metabase/driver/snowflake.clj:
```diff
@@ -54,19 +56,38 @@
+    password
+    details
+
+    private-key-path
+    (let [secret-map       (secret/db-details-prop->secret-map details "private-key")
+          private-key-file (when (some? (:value secret-map))
+                             (secret/value->file! secret-map :snowflake))]
+      (cond-> (apply dissoc details (vals (secret/get-sub-props "private-key")))
+        private-key-file (handle-conn-uri user account private-key-file)))
```
Yes. The bug that this fixes is being able to handle uploaded keys. If the `private-key-path` is empty there can still be a `private-key-value`.

metamben replied:
I meant the case when `private-key-path` is truthy, but `private-key-file` is not. Do we want to still look at `private-key-value`?

escherize replied:
I don't think we should consider `private-key-value` if `private-key-path` is truthy.

If `private-key-path` is truthy, then we should expect there to be a key at that path, and respond with a message indicating that we were unable to find one.

I don't see how it makes sense to upload a key, and give us a local path to (the same?) key. It's not possible in the UI either.

escherize committed:
update use utf-8 when encoding private-key-value

conveniently use `bytes` and `bytes?`

escherize committed:
Merge branch 'master' into fix-snowflake-private-key-issue

escherize committed:
Merge branch 'master' into fix-snowflake-private-key-issue

escherize committed:
Merge branch 'master' into fix-snowflake-private-key-issue
